### PR TITLE
Fix EventSource Schedule passes Enabled to State

### PR DESCRIPTION
**Pull request by lightpriest**
*Issue*

I stumbled upon this but noticed there's an open issue: #1271 

*Description of changes:*

In the documentation, it is stated that Enabled is passed directly to State,
and also that intrinsic functions are supported in all of the Schedule
EventSource properties.

Added the expected "passing" of the value, while maintaining the previous
behavior of `True -> ENABLED` and `False -> DISABLED`.

*Description of how you validated changes:*

Wrote missing tests for `Schedule` event source, and for this specific test.

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation
- [x] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by lightpriest** [comment]
I couldn't verify this change on a real template, how should I do this? I tried running the script under `bin` but it always failed with some different error (on packaged template, local, built, etc.) and I didn't have time to dig into each.

How do I translate a template that I have locally?

**Comment by lightpriest** [comment]
Got it to work. A fully working template renders successfully.

**Comment by lightpriest** [comment]
Any comments? Something? :)

**Comment by Schwusch** [comment]
I would like to see this PR merged. We currently have no workaround other than to look at other solutions like CDK.

File: samtranslator/model/eventsources/push.py [hunk 1]
```diff
@@ -115,7 +115,10 @@ def to_cloudformation(self, **kwargs):
 
         events_rule.ScheduleExpression = self.Schedule
         if self.Enabled is not None:
-            events_rule.State = "ENABLED" if self.Enabled else "DISABLED"
+            if isinstance(self.Enabled, bool):
+                events_rule.State = "ENABLED" if self.Enabled else "DISABLED"
+            else:
+                events_rule.State = self.Enabled
```
Comment (Schwusch) [inline samtranslator/model/eventsources/push.py:121]: Can `self.Enabled` be an [instrinsic function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) at this point?
If so, is there a way to evaluate it?

**Comment by benkehoe** [comment]
I'd also love to see this PR merged. Ran into this today and had to explode a couple schedule event sources into constituent rules, IAM roles, and Lambda permissions.

File: samtranslator/model/eventsources/push.py (hunk 1 above)
Comment (lightpriest) [reply samtranslator/model/eventsources/push.py:121]: I think that if it's passed directly to CloudFormation, it becomes an intrinsic function, isn't it so?

**Comment by lightpriest** [comment]
Anyone? 🥺

**Comment by milesgranger** [comment]
Ping as a plus one for this, quite unexpected behavior and uneconomic if we're talking about 100s of events that need to be toggled via parameters. :+1: 
Thanks for the work here @lightpriest :clap:

File: samtranslator/model/eventsources/push.py [hunk 2]
```diff
@@ -96,6 +96,7 @@ class Schedule(PushEventSource):
         "Schedule": PropertyType(True, is_str()),
         "Input": PropertyType(False, is_str()),
         "Enabled": PropertyType(False, is_type(bool)),
+        "State": PropertyType(False, is_str()),
```
Comment (hawflau) [inline samtranslator/model/eventsources/push.py:99]: Why do we want to add the `State` property? 

Essentially we just want to add intrinsic support (at least for `Fn::If` and `!Ref`). So, what we actually should do is:
- If `Enabled` is an intrinsic, just assign it to `events_rule.State`
- If `Enabled` is a string, keep it as how we handle it today (for sake of not introducing breaking change)

**Review by hawflau — CHANGES_REQUESTED** [review]

File: samtranslator/model/eventsources/push.py (hunk 2 above)
Comment (hawflau) [reply samtranslator/model/eventsources/push.py:99]: okay, looking at this [comment](https://github.com/aws/serverless-application-model/issues/1329#issuecomment-1157759679), and realizing adding intrinsic support to `Enabled` also introduces breaking change, I think adding `State` property makes sense to me now.

File: samtranslator/model/eventsources/push.py [hunk 3]
```diff
@@ -122,8 +123,18 @@ def to_cloudformation(self, **kwargs):
...
+
+        if self.State and self.Enabled:
+            raise InvalidEventException(self.relative_id, "State and Enabled Properties cannot both be specified.")
+
+        if self.State:
+            events_rule.State = self.State
+
         if self.Enabled is not None:
-            events_rule.State = "ENABLED" if self.Enabled else "DISABLED"
+            if isinstance(self.Enabled, bool):
```
Comment (hawflau) [inline samtranslator/model/eventsources/push.py:134]: If we are introducing a new property to get the behavior right with intrinsic support, why don't we just leave this as it was before?

File: samtranslator/model/eventsources/push.py [hunk 4]
```diff
@@ -122,8 +123,18 @@ def to_cloudformation(self, **kwargs):
         resources.append(events_rule)
 
         events_rule.ScheduleExpression = self.Schedule
+
+        if self.State and self.Enabled:
```
Comment (hawflau) [inline samtranslator/model/eventsources/push.py:127]: Seems like we should explicitly check `self.Enabled` is not `None`. Consider:
```
Enabled: false # bool
State: ENABLED
```

**Review by hawflau — CHANGES_REQUESTED** [review]

File: samtranslator/model/eventsources/push.py (hunk 3 above)
Comment (jfuss) [reply samtranslator/model/eventsources/push.py:134]: I meant to revert this back but looks like it got lost in me putting this task up and down. Will update.

**Review by hawflau — APPROVED** [review]
LGTM!

**Review by mildaniel — APPROVED** [review]

**Pull request merged**